src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,465 @@
+
+    void notify() { cv.notify_one(); }
+
+private:
+    void setFlag(const T & key, bool value)
+    {
+        any_need_update = any_need_update || value;
+
+        auto it = flags.find(key);
+        if (it == flags.end())
+            throw DB::Exception(ErrorCodes::LOGICAL_ERROR, "Unknown value '{}'", key);
+        it->second = value;
+    }
+
+    std::condition_variable cv;
+    std::mutex mu;
+
+    /// flag indicates that update is required
+    std::unordered_map<T, std::atomic_bool> flags;
+    std::atomic_bool any_need_update = true;
+};
+
+ClusterDiscovery::ClusterDiscovery(
+    const Poco::Util::AbstractConfiguration & config,
+    ContextPtr context_,
+    const String & config_prefix)
+    : context(Context::createCopy(context_))
+    , current_node_name(toString(ServerUUID::get()))
+    , log(&Poco::Logger::get("ClusterDiscovery"))
+{
+    LOG_DEBUG(log, "Cluster discovery is enabled");
+
+    Poco::Util::AbstractConfiguration::Keys config_keys;
+    config.keys(config_prefix, config_keys);
+
+    for (const auto & key : config_keys)
+    {
+        String prefix = config_prefix + "." + key + ".discovery";
+        if (!config.has(prefix))
+            continue;
+
+        clusters_info.emplace(
+            key,
+            ClusterInfo(
+                /* name_= */ key,
+                /* zk_root_= */ config.getString(prefix + ".path"),
+                /* port= */ context->getTCPPort(),
+                /* secure= */ config.getBool(prefix + ".secure", false),
+                /* shard_id= */ config.getUInt(prefix + ".shard", 0)
+            )
+        );
+    }
+    clusters_to_update = std::make_shared<UpdateFlags>(config_keys.begin(), config_keys.end());
+}
+
+/// List node in zookeper for cluster
+Strings ClusterDiscovery::getNodeNames(zkutil::ZooKeeperPtr & zk,
+                                       const String & zk_root,
+                                       const String & cluster_name,
+                                       int * version,
+                                       bool set_callback)
+{
+    auto watch_callback = [cluster_name, clusters_to_update=clusters_to_update](auto) { clusters_to_update->set(cluster_name); };
+
+    Coordination::Stat stat;
+    Strings nodes = zk->getChildrenWatch(getShardsListPath(zk_root), &stat, set_callback ? watch_callback : Coordination::WatchCallback{});
+    if (version)
+        *version = stat.cversion;
+    return nodes;
+}
+
+/// Reads node information from specified zookeeper nodes
+/// On error returns empty result
+ClusterDiscovery::NodesInfo ClusterDiscovery::getNodes(zkutil::ZooKeeperPtr & zk, const String & zk_root, const Strings & node_uuids)
+{
+    NodesInfo result;
+    for (const auto & node_uuid : node_uuids)
+    {
+        String payload;
+        bool ok = zk->tryGet(getShardsListPath(zk_root) / node_uuid, payload) &&
+                  NodeInfo::parse(payload, result[node_uuid]);
```

Comment:
We still does not distinguish parse error and zk error :(
Also I would rather expect an exception from `NodeInfo::parse`

Reply:
I've seen that parsing write error to log, so maybe it's ok.